lock-api/src/main/java/com/palantir/lock/watch/ConjureLockRequestMetadataUtils.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import com.palantir.atlasdb.timelock.api.ConjureChangeMetadata;
+import com.palantir.atlasdb.timelock.api.ConjureCreatedChangeMetadata;
+import com.palantir.atlasdb.timelock.api.ConjureDeletedChangeMetadata;
+import com.palantir.atlasdb.timelock.api.ConjureLockRequestMetadata;
+import com.palantir.atlasdb.timelock.api.ConjureUnchangedChangeMetadata;
+import com.palantir.atlasdb.timelock.api.ConjureUpdatedChangeMetadata;
+import com.palantir.conjure.java.lib.Bytes;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.watch.ChangeMetadata.Created;
+import com.palantir.lock.watch.ChangeMetadata.Deleted;
+import com.palantir.lock.watch.ChangeMetadata.Unchanged;
+import com.palantir.lock.watch.ChangeMetadata.Updated;
+import com.palantir.util.IndexEncodingUtils;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public final class ConjureLockRequestMetadataUtils {
+    private ConjureLockRequestMetadataUtils() {}
+
+    public static ConjureLockRequestMetadata toConjureIndexEncoded(
+            List<LockDescriptor> orderedLockDescriptors, LockRequestMetadata metadata) {
+        ChangeMetadataToConjureVisitor toConjureVisitor = new ChangeMetadataToConjureVisitor();
+        Map<Integer, ConjureChangeMetadata> indexEncodedConjureChangeMetadata = IndexEncodingUtils.encode(
+                orderedLockDescriptors,
+                metadata.lockDescriptorToChangeMetadata(),
+                changeMetadata -> changeMetadata.accept(toConjureVisitor));
+        return ConjureLockRequestMetadata.of(indexEncodedConjureChangeMetadata);
+    }
+
+    public static LockRequestMetadata fromConjureIndexEncoded(
+            List<LockDescriptor> orderedLockDescriptors, ConjureLockRequestMetadata indexEncodedConjureMetadata) {
+        ChangeMetadataFromConjureVisitor fromConjureVisitor = new ChangeMetadataFromConjureVisitor();
+        Map<LockDescriptor, ChangeMetadata> changeMetadata = IndexEncodingUtils.decode(
+                orderedLockDescriptors,
+                indexEncodedConjureMetadata.getIndexToChangeMetadata(),
+                conjureChangeMetadata -> conjureChangeMetadata.accept(fromConjureVisitor));
+        // visitUnknown() will return null
+        changeMetadata.values().removeIf(Objects::isNull);
```

Comment:
Do we want to do this here? This is unnecessary in most cases and propagating null here does not seem that bad

Reply:
Let's use Optionals

Reply:
Also, why are these maps mutable? We should make them immutable!

Reply:
immutability has to be enforced in the index encoding layer, I will address this in the corresponding PR